ledger/ledger-utxo/src/main/kotlin/net/corda/v5/ledger/utxo/UtxoLedgerService.kt

```diff
@@ -120,4 +121,59 @@ interface UtxoLedgerService {
         session: FlowSession,
         validator: UtxoTransactionValidator
     ): UtxoSignedTransaction
-}
+
+    /**
+     * Sends a transaction builder to another session, then it waits for other side to propose transaction builder components,
+     * then applies the proposed components to a copy of the original builder and returns that new builder.
+     *
+     * It supports similar workflows:
+     * Initiator:
+     *
+     * val updatedTxBuilder = utxoLedgerService.sendAndReceiveTransactionBuilder(txBuilder, session)
+     *
+     * The proposal one's notary and time window will get discarded and the original will be kept if both the original and
+     * the proposal have these components set. Also, all duplications will be discarded.
+     *
+     * Receiver:
+     *
+     * val proposedTxBuilder = utxoLedgerService.receiveTransactionBuilder(session)
+     * proposedTxBuilder.add...(...)
+     * proposedTxBuilder.add...(...)
+     * proposedTxBuilder.add...(...)
+     * utxoLedgerService.replyTransactionBuilderProposal(proposedTxBuilder, session)
+     *
+     * @param transactionBuilder The [UtxoTransactionBuilder] to send.
+     * @param session The receiver [FlowSession].
+     *
+     * @return A new merged builder of the original and proposed components.
+     */
+    @Suspendable
+    fun sendAndReceiveTransactionBuilder(
+        transactionBuilder: UtxoTransactionBuilder,
+        session: FlowSession,
+    ): UtxoTransactionBuilder
+
+    /**
+     * Receives a transaction builder from another session.
+     *
+     * @param session The [FlowSession] to receive the [UtxoTransactionBuilder] from.
+     */
+    @Suspendable
+    fun receiveTransactionBuilder(
+        session: FlowSession,
+    ): UtxoTransactionBuilder
+
+    /**
+     * Sends the differences of transaction builders to another session with all dependent back chains.
+     * It works only with [receiveTransactionBuilder] created [UtxoTransactionBuilder]s which track the differences internally.
+     * If it is called with anything else, it throws [InvalidParameterException].
+     *
+     * @param transactionBuilder The [UtxoTransactionBuilder] to send.
+     * @param session The receiver [FlowSession].
+     */
+    @Suspendable
+    fun replyTransactionBuilderProposal(
```

Comment:
I would prefer this was called `sendUpdatedTransactionBuilder`, but happy to see what others think.